Describe note discovery note tags (#713)
0xPolygonMiden/miden-base: docs/architecture/notes.md

* P2ID and P2IDR scripts are used to send assets to a specific account ID. The scripts check at note consumption if the executing account ID equals the account ID that was set by the note creator as note inputs. The P2IDR script is reclaimable and thus after a certain block height can also be consumed by the sender itself.
* SWAP script is a simple way to swap assets. It adds an asset from the note into the consumer's vault and creates a new note consumable by the first note's issuer containing the requested asset.

> ??? note "Example note script pay to ID (P2ID)"

Reviewer: make sure you check it renders correctly locally

!!! info
    To compute a note's hash, we do not need to know the note's `serial_num`. Knowing the hash of the `serial_num` (as well as `script_hash`, `input_hash` and `note_vault`) is also sufficient. We compute the hash of `serial_num` as `hash(serial_num, [0; 4])` to simplify processing within the VM._

> ## Note discovery (note tags)

Author: Jup, I also have the feeling that the entire note chapter is too long already. I will add the example for now and we should do a bigger refactor for the chapter and add several sub-sections.

> ## Note discovery (note tags)

Author: In general, I think, what is missing to make those really great docs:

- the ppl who write / design the code should also write the section (they know exactly why a certain design was chosen)
- Miden cannot be used at the moment, but the docs should describe how to use Miden. It will get better, once there is the compiler. Then we can describe how to use a note tag using a specific example.

> Note discovery describes the process by which Miden clients find notes they want to consume. Miden clients can query the Miden node for notes carrying a certain note tag. Note tags are best-effort filters for notes registered on the network. They are lightweight values used to speed up queries. Clients can follow tags for specific use cases, such as swap scripts, or agree with a group of users on certain tags for their specific needs. Tags are also used by the operator to identify notes intended for network execution and include the corresponding information on how to execute them.

Reviewer: ```suggestion
Note discovery describes the process by which Miden clients find notes they want to consume. Miden clients can query the Miden node for notes carrying a certain note tag. Note tags are best-effort filters for notes registered on the network. They are lightweight values used to speed up queries. Clients can follow tags for specific use cases, such as swap scripts, or user-created custom tags. Tags are also used by the operator to identify notes intended for network execution and include the corresponding information on how to execute them.
```

Reviewer: This sentence was problematic. It reads like "clients agreeing with users" which is strange. The main verb is "follow" which you don't need to repeat.

```arduino
0b009f4adc47857e2f6

> ```

Reviewer 2: A couple of other things probably worth mentioning here:
- Note tags are 32-bit values (we say "lightweight" above - but I think it may be good to make it concrete).
- Note tags are a part of note metadata which is always public (not sure if we have a separate section about note metadata somewhere).

Author: done. 

We mention note metadata and explain that this is always public. But there is no separate section as we changed it to explain the note lifecycle rather than the note concepts.

> The example note tag above indicates that the network operator (Miden node) executes the note against the account with the ID `0x09f4adc47857e2f6`. In this case, the note and the account against which it gets executed must be `public`.

Reviewer: ```suggestion
The example note tag above indicates that the network operator (Miden node) executes the note against the account with ID `0x09f4adc47857e2f6`. In this case, the note and the account against which it gets executed must be `public`.
```

The two most signification bits of the note tag have the following interpretation:

| Prefix | Execution hint | Target   | Allowed note type |
| ------ | :------------: | :------: | :----------------:|
| `0b00` | Network        | Specific | NoteType::Public  |
| `0b01` | Network        | Use case | NoteType::Public  |
| `0b10` | Local          | Any      | NoteType::Public  |
| `0b11` | Local          | Any      | Any               |

> - Execution hint: Set to `Network` for network transactions. These notes will be validated and, if possible, consumed in a network transaction.

Reviewer: ```suggestion
- Execution hint: Set to `Network` for network transactions. These notes are validated and, if possible, consumed in a network transaction.
```

- Target: Describes how to interpret the bits in the note tag. For tags with a specific target, the rest of the tag is interpreted as an `account_id`. For use case values, the meaning of the rest of the tag is not specified by the protocol and can be used by applications built on top of the rollup.
- Note type: Describes the note's storage mode, either `public` or `private`.

> The following 30 bits can represent anything; from account IDs to use cases or any custom logic agreed upon.

Reviewer: ```suggestion
The following 30 bits can represent anything; from account IDs to use cases or any custom logic.
```

Reviewer: the assumption is custom logic has already been agreed upon, you don't need to state it

!!! info
    To compute a note's hash, we do not need to know the note's `serial_num`. Knowing the hash of the `serial_num` (as well as `script_hash`, `input_hash` and `note_vault`) is also sufficient. We compute the hash of `serial_num` as `hash(serial_num, [0; 4])` to simplify processing within the VM._

> ## Note discovery (Note Tags)

Reviewer: ```suggestion
## Note discovery (note tags)
```

> Note discovery describes the process of Miden clients finding notes they want to consume. Miden clients can query the Miden node for notes carrying a certain note tag. Note tag`s are best effort filters for notes registered with the network. They are light-weight values used to speed up queries. Clients can follow tags of a certain use case, e.g. swap scripts, or agree upon a group of users on certain tags for their specific use case. Tags are also used by the operator to identify notes intended for network execution including the corresponding information on how to execute.

Reviewer: ```suggestion
Note discovery describes the process of Miden clients finding notes they want to consume. Miden clients can query the Miden node for notes carrying a certain note tag. Note tag`s are best effort filters for notes registered on the network. They are light-weight values used to speed up queries. Clients can follow tags for a certain use case, e.g. swap scripts, or agree upon a group of users of certain tags for their specific use case. Tags are also used by the operator to identify notes intended for network execution and include the corresponding information on how to execute.
```

Reviewer: I'm not sure what this means "or agree upon a group of users of certain tags for their specific use case"...

Author: I change it to 

```
Clients can follow tags for a certain use case, e.g. swap scripts, or agree with a group of users on certain tags for their specific use case.
```

```arduino
0b009f4adc47857e2f6
```

> This example note tag indicates that the network operator (Miden node) shall execute the note against the account with the ID `0x09f4adc47857e2f6`. In this case, the note and the account against it gets executed must be `public`.

Reviewer: ```suggestion
The example note tag below indicates that the network operator (Miden node) executes the note against the account with ID `0x09f4adc47857e2f6`. In this case, the note and the account must be `public`.
```

The two most signification bits of the note tag have the following interpretation:

| Prefix | Execution hint | Target   | Allowed note type |
| ------ | :------------: | :------: | :----------------:|
| `0b00` | Network        | Specific | NoteType::Public  |
| `0b01` | Network        | Use case | NoteType::Public  |
| `0b10` | Local          | Any      | NoteType::Public  |
| `0b11` | Local          | Any      | Any               |

Where:

- Execution hint is set to `Network` for network transactions. These notes will be further validated and if possible consumed in a network transaction.
- Target describes how to further interpret the bits in the note tag. For tags with a specific target, the rest of the tag is interpreted as an `account_id`. For use case values, the meaning of the rest of the tag is not specified by the protocol and can be used by applications built on top of the rollup.
- Note type describes the note's storage mode, either public or private.

> The following 30 bits can represent anything, from Account IDs to use cases or any custom logic agreed upon.

Reviewer: ```suggestion
The following 30 bits can represent anything; from Account IDs to use cases or any custom logic agreed upon.
```

> Using note tags is a compromise between privacy and latency. If a user would query the operator using the note ID, the operator would learn in which note a specific user is interested in. On the other side, if a user would always download all registered notes and filter locally, it would be quite inefficient. Using tags, users can decide their privacy parameters by narrowing or broadening their note tag schemes.

Reviewer: ```suggestion
Using note tags is a compromise between privacy and latency. If a user queries the operator using the note ID, the operator learns which note a specific user is interested in. Alternatively, if a user always downloads all registered notes and filters locally, it is quite inefficient. Using tags, users can customize privacy parameters by narrowing or broadening their note tag schemes.
```

| `0b10` | Local          | Any      | NoteType::Public  |
| `0b11` | Local          | Any      | Any               |

> - Where:

Reviewer: you don't need this where, it's strange ..

> - Execution Hint: Set to `Network` for network transactions. These notes will be further validated and, if possible, consumed in a network transaction.

Reviewer: hint was lower case in the table i think ...

Reviewer: instead of "will be" -> "are"

> - Target: Describes how to further interpret the bits in the note tag. For tags with a specific target, the rest of the tag is interpreted as an `account_id`. For use case values, the meaning of the rest of the tag is not specified by the protocol and can be used by applications built on top of the rollup.

Reviewer: remove "further"

> - Note type describes the note's storage mode, either `public` or `private`.

Reviewer: Add the colon: Capital letter .. for consistency

> The following 30 bits can represent anything—from Account IDs to use cases or any custom logic agreed upon.

Reviewer: use a ; here (dash is incorrect)

Reviewer: account should be small A
